HW2/P5/physics.pyx

```diff
@@ -69,17 +73,28 @@ cdef void sub_update(FLOAT[:, ::1] XY,
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
+
+    # get grid index of XY1
+    gridx = int(XY1[0]/grid_spacing)
+    gridy = int(XY1[1]/grid_spacing)
+    for x in range(max(0, gridx-search_size), min(Grid.shape[0], gridx+search_size+1)):
```

Comment:
Instead of checking if i < j, you can make sure to only look three steps ahead of x (not x-3 to x+3), and same for y. That way, you know you never double count, and you run fewer computations